### DISCUSSION TITLE
check_hdr_str() error while reading in output from ascii2nc

### DISCUSSION THREAD

**Discussion by davis-wesley-12345** [dtcenter/METplus#1856] [Configuration] [answered]
I've used ascii2nc to produce a file full of observations to compare an ensemble_stat run against.

I'm running into an issue where ensemble stat returns the following error:

`WARNING: check_hdr_str() -> null string!
`

An ncdump -h of the .nc file returns the following:

```
        char hdr_typ_table(nhdr_typ, mxstr2) ;
                hdr_typ_table:long_name = "message type" ;
```

Looking for the specific header type that should be found, I do an ncdump and find the following:

```
 hdr_typ_table =
  "169" ;
```

So this tells me that I should have that value in my message_type_group_map dictionary. Here's what I have:

```
message_type   = [ "MIGHTI_U" ];
message_type_group_map = [
   { key = "SURFACE"; val = "ADPSFC,SFCSHP,MSONET";               },
   { key = "ANYAIR";  val = "AIRCAR,AIRCFT";                      },
   { key = "ANYSFC";  val = "ADPSFC,SFCSHP,ADPUPA,PROFLR,MSONET"; },
   { key = "ONLYSF";  val = "ADPSFC,SFCSHP";                      },
   { key = "MIGHTI_U";  val = "169"; }
];
```

Given this and the data I just uploaded, can you give me some direction on how I need to fill out my message_type_group map to make this ensemble_stat run work?

FYI I'm working off of v9.1

Thank you!

**Comment by j-opatz** [comment]
Hello,

And thank you for your question. It sounds like what you're trying to accomplish is have Ensemble-Stat process the ASCII2NC output file, which currently contains a variable with the message type "169" as indicated by your `hdr_typ_table`.

The `message_type_group_map` dictionary is to allow users quick grouping of point observation message types for verification; this will not allow users to create a value-key pairing as you're attempting to do. So no adjustments should be made to your `message_type_group_map` dictionary.

Instead, you will need to utilize the `message_type` list. By setting this to the message type indicated in your input file, Ensemble-Stat should understand what it's supposed to analyze. So the following should be all you need to get your configuration file to work:

```

message_type    =  [  "169"  ];

```

Please let us know if this change helps with the issue or if further assistance is needed.

**Comment by davis-wesley-12345** [comment]
Unfortunately I'm getting the same error after changing message_type to [ "169" ]

message_type   = [ "169" ];
message_type_group_map = [
   { key = "SURFACE"; val = "ADPSFC,SFCSHP,MSONET";               },
   { key = "ANYAIR";  val = "AIRCAR,AIRCFT";                      },
   { key = "ANYSFC";  val = "ADPSFC,SFCSHP,ADPUPA,PROFLR,MSONET"; },
   { key = "ONLYSF";  val = "ADPSFC,SFCSHP";                      }
];

```
DEBUG 2: Processing mean file: /p/work1/daviswj/UV-126/uv/scripts/LETKF_ensemble/OUT_ENS_L_T119L160_99439994f_0401IC_NS/uuwind_pre_000.10_006_2020042612/ensemble_stat_20200426_120000V_ens.nc
DEBUG 4: Generated mean field: uuwind_pre_000.10_ENS_MEAN
DEBUG 4: Met2dDataFileFactory::new_met_2d_data_file() -> created new Met2dDataFile object of type "FileType_NcMet".
DEBUG 4:
DEBUG 4: Latitude/Longitude Grid Data:
DEBUG 4:      lat_ll: -90
DEBUG 4:      lon_ll: -0
DEBUG 4:   delta_lat: 1
DEBUG 4:   delta_lon: 1
DEBUG 4:        Nlat: 181
DEBUG 4:        Nlon: 360
DEBUG 4:
DEBUG 7: check_nc_data_2d() count: valid=65160, zero=0, missing=0, 65160 == 65160
DEBUG 7: MetNcFile::data() const -> took 0.02 seconds to read NetCDF data
DEBUG 7: MetNcFile::data() const -> took 0 seconds to fill data plane
DEBUG 2: For uuwind_pre_000.10_ENS_MEAN(*,*) found 1 forecast levels.
DEBUG 2:
DEBUG 2: --------------------------------------------------------------------------------
DEBUG 2:
WARNING:
WARNING: check_hdr_str() -> null string!
WARNING:
```

**Reply by j-opatz** [reply]
That's unfortunate to hear,

Would you be able to send along 2 pieces of data for us to test:
- The netCDF input file
- A copy of your configuration file

With those in hand, I'll take a look and see if I can recreate the issue and find a way around it. Details for submitting data are here: https://github.com/dtcenter/METplus/discussions/954

**Reply by davis-wesley-12345** [reply]
No problem, just took care of that. 

fcst -> ensemble_stat_20200426_120000V_ens.nc
config -> Standard_PointStatConfig_NAVGEM_2020042612
obs -> innovar_norad_1_2020042612.nc

**Comment by j-opatz** [comment, marked as answer]
I was able to take a look at the data you sent over and found a few things that may need adjustment in your configuration file before it successfully runs:

- When I first ran your configuration file as-is, Point-Stat actually ran successfully. I needed to update a few paths to be relevant to my working area (namely, the `tmp` path and `poly` masks) and the MET version number (I tested against the latest develop branch and an older 10.0.1 branch), but there were no ```check_hdr_str() -> null string! ```warnings, only a depreciated variable warning. We currently don't have a local 9.1 branch to test against, so there _is_ a small chance that the older MET version is encountering something that was corrected in a later release. I noticed in your printout of the issue that it's listed as a `WARNING` and not an `ERROR`, so most likely the version difference shouldn't matter. However, if the following configuration changes don't alleviate the issue, then I would say a newer MET version will likely resolve any lingering issues.

- In your configuration file observation field dictionary, you've listed the name as `UGRD`. But when Point-Stat completed running, all of the observations were being thrown out for a naming issue:
 ```
DEBUG 2: Number of matched pairs   = 0
DEBUG 2: Observations processed    = 14722
DEBUG 2: Rejected: station id      = 0
DEBUG 2: Rejected: obs var name    = 14722
```

 A look at the `obs_var` variable for innovar_norad_1_2020042612.nc shows why:
```
ncdump -v obs_var innovar_norad_1_2020042612.nc 
netcdf innovar_norad_1_2020042612 {
data:

 obs_var =
  "UWND" ;
}
```
 so you'll need to change the observation field name
```
     name       = "UWND";
```
- After that, I reran the case and found that Point-Stat was tossing out the observations due to the time difference between model valid time and observation valid time:

```
DEBUG 2: Number of matched pairs   = 0
DEBUG 2: Observations processed    = 14722
DEBUG 2: Rejected: station id      = 0
DEBUG 2: Rejected: obs var name    = 0
DEBUG 2: Rejected: valid time      = 14722
```
This correction will need to be personally adjusted by you, but for testing purposes I put the observation window incredibly large, to ensure none would be rejected. That adjustment looked like this:
```
obs_window = {
   beg = -54000;
   end =  54000;
}
```
This large window (+/-15 hours) now allows for all of the observations to be compared to the model, with a rerun of the use case showing the following rejections (and matched pairs):
```
DEBUG 3: Number of matched pairs   = 647
DEBUG 3: Observations processed    = 14722
DEBUG 3: Rejected: station id      = 0
DEBUG 3: Rejected: obs var name    = 0
DEBUG 3: Rejected: valid time      = 0
DEBUG 3: Rejected: bad obs value   = 0
DEBUG 3: Rejected: off the grid    = 0
DEBUG 3: Rejected: topography      = 0
DEBUG 3: Rejected: level mismatch  = 14075
DEBUG 3: Rejected: quality marker  = 0
DEBUG 3: Rejected: message type    = 0
DEBUG 3: Rejected: masking region  = 0
DEBUG 3: Rejected: bad fcst value  = 0
DEBUG 3: Rejected: bad climo mean  = 0
DEBUG 3: Rejected: bad climo stdev = 0
DEBUG 3: Rejected: mpr filter      = 0
DEBUG 3: Rejected: duplicates      = 0
DEBUG 2: Computing Scalar Partial Sums and Continuous Statistics.
```
Looking at the pressure thresholds you're using on the point data (P1.1-.9) and the general levels of the observations, this seems correct. 

So I hope those two edits (field name and a wider observation window) will help yield some results. If not, you'll need to obtain a newer release of MET.

**Reply by davis-wesley-12345** [reply]
I ended up figuring it out thanks to your tips. I guess I was just confused that the ensemble_stat netcdf output needed to be read in to MET using the netcdf rules and the ascii2nc output needed to be read in using the GRIB rules.

```
fcst = {
  field =
  [
   { name = "uuwind_pre_0001.0_ENS_MEAN";
     level = "(*,*)"; }
  ];
}

obs = {
  field =
  [ { name = "UWND";
      level = "P.9-1.1";}
  ];
}

```

**Comment by davis-wesley-12345** [comment]
can you maybe post the config file you came up with?

**Reply by j-opatz** [reply]
Sure can, I've attached my markup version here. You'll note the changes I had to make in order for it to run on our machines (version number, commenting out the poly files, changing the tmp directory), in addition to the changes I suggested above that will let you start having matched pairs.
[Standard_PointStatConfig_NAVGEM_2020042612_markup.txt](https://github.com/dtcenter/METplus/files/9840735/Standard_PointStatConfig_NAVGEM_2020042612_markup.txt)
